src/logger.ts

```diff
@@ -0,0 +1,55 @@
+/* eslint-disable @typescript-eslint/no-empty-function */
+import { Logger } from './types';
+
+const LOG_LEVELS: Array<keyof Logger> = ['error', 'warn', 'info', 'debug'];
+export class DefaultLogger implements Logger {
+  error(...args: unknown[]): void {
+    console.error(...args);
+  }
+  warn(...args: unknown[]): void {
+    console.warn(...args);
+  }
+  info(): void {}
+  debug(): void {}
+}
+
+export class SafeLogger implements Logger {
+  private readonly logger: Logger;
+  private readonly fallbackLogger = new DefaultLogger();
+
+  constructor(logger: Logger) {
+    try {
+      for (const level of LOG_LEVELS) {
+        if (!logger[level] || typeof logger[level] !== 'function') {
+          throw new Error(`The provided logger is missing the ${level} method.`);
+        }
+      }
+      this.logger = logger;
+    } catch (err) {
+      console.error(err);
+      console.warn('Falling back to the default logger.');
+      this.logger = this.fallbackLogger;
+    }
+  }
+
+  error(...args: unknown[]): void {
+    this.log('error', ...args);
+  }
+  warn(...args: unknown[]): void {
+    this.log('warn', ...args);
+  }
+  info(...args: unknown[]): void {
+    this.log('info', ...args);
+  }
+  debug(...args: unknown[]): void {
+    this.log('debug', ...args);
+  }
+
+  private log(level: keyof Logger, ...args: unknown[]) {
+    try {
+      this.logger[level](...args);
```

Comment:
I actually think you  should call this with `.apply` or `.call` and pass the logger as the `thisArg`. If the logger method called any members, they would be undefined.

Reply:
Nevermind, I think this is fine since you're still calling this as a method, not a function ref.